pkg/proc/native/exec_darwin.c

```diff
@@ -5,110 +5,76 @@
-}
+#ifndef POSIX_SPAWN_DISABLE_ASLR
+#define POSIX_SPAWN_DISABLE_ASLR 0x0100
+#endif
 
 int
-fork_exec(char *argv0, char **argv, int size,
+spawn(char *argv0, char **argv, int size,
 		char *wd,
 		task_t *task,
 		mach_port_t *port_set,
 		mach_port_t *exception_port,
 		mach_port_t *notification_port)
 {
-	// Since we're using mach exceptions instead of signals,
-	// we need to coordinate between parent and child via pipes
-	// to ensure that the parent has set the exception ports on
-	// the child task before it execs.
-	int fd[2];
-	if (close_exec_pipe(fd) < 0) return -1;
-
-	// Create another pipe to signal the parent on exec.
-	int efd[2];
-	if (close_exec_pipe(efd) < 0) return -1;
-
-	kern_return_t kret;
-	pid_t pid = fork();
-	if (pid > 0) {
-		// In parent.
-		close(fd[0]);
-		close(efd[1]);
-		kret = acquire_mach_task(pid, task, port_set, exception_port, notification_port);
-		if (kret != KERN_SUCCESS) return -1;
-
-		char msg = 'c';
-		write(fd[1], &msg, 1);
-		close(fd[1]);
-
-		char w;
-		size_t n = read(efd[0], &w, 1);
-		close(efd[0]);
-		if (n != 0) {
-			// Child died, reap it.
-			waitpid(pid, NULL, 0);
-			return -1;
-		}
-		return pid;
-	}
+	kern_return_t kret = 0;
 
-	// Fork succeeded, we are in the child.
-	int pret, cret;
-	char sig;
+    posix_spawn_file_actions_t file_actions;
+    posix_spawnattr_t attributes;
 
-	close(fd[1]);
-	read(fd[0], &sig, 1);
-	close(fd[0]);
+	// TODO: check error
+    posix_spawnattr_init(&attributes);
 
-	// Create a new process group.
-	if (setpgid(0, 0) < 0) {
-		perror("setpgid");
-		exit(1);
-	}
+    sigset_t no_signals;
+    sigset_t all_signals;
+    sigemptyset(&no_signals);
+    sigfillset(&all_signals);
 
-	// Set errno to zero before a call to ptrace.
-	// It is documented that ptrace can return -1 even
-	// for successful calls.
-	errno = 0;
-	pret = ptrace(PT_TRACE_ME, 0, 0, 0);
-	if (pret != 0 && errno != 0) {
-		perror("ptrace");
-		exit(1);
-	}
+    posix_spawnattr_setsigmask(&attributes, &no_signals);
+    posix_spawnattr_setsigdefault(&attributes, &all_signals);
 
-	// Change working directory if wd is not empty.
-	if (wd && wd[0]) {
-		errno = 0;
-		cret = chdir(wd);
-		if (cret != 0 && errno != 0) {
-			char *error_msg;
-			asprintf(&error_msg, "%s '%s'", "chdir", wd);
-			perror(error_msg);
-			exit(1);
-		}
-	}
+    // Change working directory if wd is not empty.
+    if (wd && wd[0]) {
+        errno = 0;
+        int ret = chdir(wd);
 
-	errno = 0;
-	pret = ptrace(PT_SIGEXC, 0, 0, 0);
-	if (pret != 0 && errno != 0) {
-		perror("ptrace");
-		exit(1);
-	}
+        if (ret != 0 && errno != 0) {
+            char *error_msg;
+            asprintf(&error_msg, "%s '%s'", "chdir", wd);
+            perror(error_msg);
+            exit(1);
+        }
+    }
 
-	sleep(1);
+    short flags = POSIX_SPAWN_START_SUSPENDED | POSIX_SPAWN_SETSIGDEF |
+          POSIX_SPAWN_SETSIGMASK;
+
+#if defined(__arm64__)
+	// Force ASLR disable on arm64. This might not be needed if a non-PIE
```

Comment:
We have a flag to control disabling ASLR, why must we force it here?

Reply:
This is a somewhat weird behaviour that occurs on darwin/arm64. Even with `buildmode=exe`, it seems that always a PIE exe is built, so we need to disable ASLR to counter it. Another, properly better way would be to use the flag and force the flag on darwin/arm64.

Reply:
I am running into a problem using the `disableASLR` flag. While it is easy to use it for the regular build, I have no way to set it during the tests. Should I extend the tests to pass it through, similarly to the backend - but this will take a lot of adjustments, or somehow force it on another level?

Reply:
Hm, yeah enabling this for tests can be tricky. You would have to modify `withTestProcessArgs` to accept launch flags. I don't know if this is worth it for now. If darwin/arm64 simply won't work without it disabled, I'm ok with a heavier-handed approach like this, but I wanted to call some attention to it first.

Reply:
I have decided to force it in the `proc_darwin.go` instead of burying deep down in the C code. Probably the lesser of the evils.